lib/puppet/functions/map.rb

```diff
@@ -1,33 +1,68 @@
-# Applies a parameterized block to each element in a sequence of entries from the first
-# argument and returns an array with the result of each invocation of the parameterized block.
+# Applies a [lambda](http://docs.puppetlabs.com/puppet/latest/reference/lang_lambdas.html)
+# to every value in a data structure and returns an array containing the results.
 #
-# This function takes two mandatory arguments: the first should be an Array, Hash, or of Enumerable type
-# (integer, Integer range, or String), and the second a parameterized block as produced by the puppet syntax:
+# This function takes two mandatory arguments, in this order:
 #
-#       $a.map |$x| { ... }
-#       map($a) |$x| { ... }
+# 1. An array or hash the function will iterate over.
+# 2. A lambda, which the function calls for each element in the first argument. It can
+# request one or two parameters.
 #
-# When the first argument `$a` is an Array or of enumerable type, the block is called with each entry in turn.
-# When the first argument is a hash the entry is an array with `[key, value]`.
+# @example Using the `map` function
 #
-# @example Using map with two arguments
+# `$transformed_data = $data.map |$parameter| { <PUPPET CODE BLOCK> }`
 #
-#      # Turns hash into array of values
-#      $a.map |$x|{ $x[1] }
+# or
 #
-#      # Turns hash into array of keys
-#      $a.map |$x| { $x[0] }
+# `$transformed_data = map($data) |$parameter| { <PUPPET CODE BLOCK> }`
 #
-# When using a block with 2 parameters, the element's index (starting from 0) for an array, and the key for a hash
-# is given to the block's first parameter, and the value is given to the block's second parameter.args.
+# When the first argument (`$data` in the above example) is an array, Puppet passes each
+# value in turn to the lambda.
 #
-# @example Using map with two arguments
+# @example Using the `map` function with an array and a one-parameter lambda
 #
-#      # Turns hash into array of values
-#      $a.map |$key,$val|{ $val }
+# ~~~ puppet
+# # For the array $data, return an array containing the values
+# $data = [1,2,3]
+# $transformed_data = $data.map |$items| { $items }
+# # $transformed_data contains [1,2,3]
```

Comment:
This example could be improved by actually transforming the data somehow. Like, `$data.map |$item| { $item * 10 }` (produces `[10, 20, 30]`).